flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceWithWatermarkExtractor.java

```diff
@@ -0,0 +1,434 @@
+                .streaming(true)
+                .monitorInterval(Duration.ofMillis(10))
+                .streamingStartingStrategy(StreamingStartingStrategy.TABLE_SCAN_THEN_INCREMENTAL)
+                .build(),
+            WatermarkStrategy.<RowData>noWatermarks()
+                .withWatermarkAlignment("iceberg", Duration.ofMinutes(20), Duration.ofMillis(10)),
+            SOURCE_NAME,
+            TypeInformation.of(RowData.class));
+
+    try (CloseableIterator<RowData> resultIterator = stream.collectAsync()) {
+      JobClient jobClient = env.executeAsync("Iceberg Source Throttling Test");
+
+      // Check that the read the non-blocked data
+      // The first RECORD_NUM_FOR_2_SPLITS should be read
+      // 1 or more from the runaway reader should be arrived depending on thread scheduling
+      waitForRecords(resultIterator, RECORD_NUM_FOR_2_SPLITS + 1);
+
+      // Get the drift metric, wait for it to be created and reach the expected state
+      // (100 min - 20 min - 0 min)
+      // Also this validates that the WatermarkAlignment is working
+      Awaitility.await()
+          .atMost(120, TimeUnit.SECONDS)
+          .until(
+              () ->
+                  findAlignmentDriftMetric(jobClient.getJobID(), TimeUnit.MINUTES.toMillis(80))
+                      .isPresent());
+      Gauge<Long> drift =
+          findAlignmentDriftMetric(jobClient.getJobID(), TimeUnit.MINUTES.toMillis(80)).get();
+
+      // Add some old records with 2 splits, so even if the blocked gets one split, the other reader
+      // one gets one as well
+      List<Record> newBatch1 =
+          ImmutableList.of(
+              generateRecord(15, "file_3-recordTs_15"),
+              generateRecord(16, "file_3-recordTs_16"),
+              generateRecord(17, "file_3-recordTs_17"));
+      List<Record> newBatch2 =
+          ImmutableList.of(
+              generateRecord(15, "file_4-recordTs_15"),
+              generateRecord(16, "file_4-recordTs_16"),
+              generateRecord(17, "file_4-recordTs_17"));
+      dataAppender.appendToTable(
+          dataAppender.writeFile(newBatch1), dataAppender.writeFile(newBatch2));
+      // The records received will highly depend on scheduling
+      // We minimally get 3 records from the non-blocked reader
+      // We might get 1 record from the blocked reader (as part of the previous batch -
+      // file_1-recordTs_100)
+      // We might get 3 records form the non-blocked reader if it gets both new splits
+      waitForRecords(resultIterator, 3);
+
+      // Get the drift metric, wait for it to be created and reach the expected state (100 min - 20
+      // min - 15 min)
+      Awaitility.await()
+          .atMost(120, TimeUnit.SECONDS)
+          .until(() -> drift.getValue() == TimeUnit.MINUTES.toMillis(65));
+
+      // Add some new records which should unblock the throttled reader
+      batch =
+          ImmutableList.of(
+              generateRecord(110, "file_5-recordTs_110"),
```

Comment:
nit: maybe the watermark can be 90 and 91?

Reply:
Done